week7/cooper_client_react/02_add_login.md

```diff
@@ -0,0 +1,418 @@
+
+Run the application to make sure that this works and that the visibility of the login form is toggled.
+
+If you run the feature test again, you will see that the change we made has not changed the error message fromthe one we got before. We need to add some logic to our component in order to make the login flow work. 
+
+We are going to add a method called `onLogin` that is going to authenticate the user against the backend system. 
+
+First, you need to add this method to the App component:
+
+```js
+async onLogin(e) {
+  e.preventDefault();
+  let resp = await authenticate(this.state.email, this.state.password)
+  if (resp.authenticated === true) {
+    this.setState({ authenticated: true });
+  } else {
+    this.setState({ message: resp.message, renderLoginForm: false })
+  }
+}
+```
+As you can see we have added some more objects to out state. We need to add default values to the constructor.
+
+```js
+  constructor(props) {
+    super(props);
+    this.state = {
+      distance: '',
+      gender: 'female',
+      age: '',
+      renderLoginForm: false,
+      authenticated: false,
+      email: '',
+      password: '',
+      message: ''
+    }
+  }
+```
+We also need to add some props to the rendering of the `LoginFrom` component.
+
+```js
+<LoginForm
+  loginHandler={this.onLogin.bind(this)}
+  inputChangeHandler={this.onChange.bind(this)}
+/>
+```
+
+What we do here is that every time we change the input fields in the `LoginForm` component we will update the state of email and password. We also add so that when the submit button is clicked, we run the `onLogin` method here in the `App` component. To make this work we need to make some changes in the `LoginForm` component.'
+
+```js
+import React from 'react';
+
+const LoginForm = (props) => {
+  return (
+    <form>
+      <div>
+        <label >Email</label>
+        <input id="email" onChange={props.inputChangeHandler}></input>
+      </div>
+
+      <div>
+        <label>Password</label>
+        <input id="password" onChange={props.inputChangeHandler}></input>
+      </div>
+      <button onClick={(e) => props.loginHandler(e)} id="submit">Submit</button>
+    </form>
+  )
+}
+
+export default LoginForm;
+```
+If you run the test now, all of them will fail. The app complains about how `authenticate` in the `App` component is not defined. 
+
+Like we did with the `CooperCalculator`, we will extract the login function to a separate module. That is what we are trying to call on with `authenticate` in the `onLogin` function. 
+
+We will make a call to our backend API over the network and will need to add a library to be able to perform that call. The library we will use is Axios. You install it like this:
```

Comment:
Perhaps add a link to the axios library?

Reply:
done